### ISSUE TITLE
Deterministic dict order

### ISSUE THREAD

**Issue by jsignell** [dask/dask#7911]
@zzhengnan has been working on unskipping doctests and as part of that work we realized that the order of many of the dicts in dask is not deterministic. Obviously this matters from the doctests perspective (https://github.com/dask/dask/pull/7874#discussion_r669227208) but it's not clear to me if it would be a generally useful thing to have predictable dict ordering. Here's an example pulled from the highlevelgraph docstring.

**Minimal Complete Verifiable Example**:

```python
>>> import dask
>>> import dask.dataframe as dd
>>> dask.datasets.timeseries(freq="1H", start="2021-07-01", end="2021-07-05").to_csv("myfile.*.csv")
>>> df = df.read_csv("myfile.*.csv")
>>> df = df[df.name == "Alice"]
>>> graph = df.__dask_graph__()
>>> graph
HighLevelGraph with 4 layers.
<dask.highlevelgraph.HighLevelGraph object at 0x7f4456afec70>
 0. read-csv-136524d4e292271449abaac2e2d5db0f
 1. getitem-063bc8eb2c47ea9536bbd7059706a728
 2. eq-55ca3c29ae249bb1a1ab9eb42332a747
 3. getitem-56f4e42f7c46ae412827cce188b37184

>>> graph.layers
{'getitem-56f4e42f7c46ae412827cce188b37184': Blockwise<(('read-csv-136524d4e292271449abaac2e2d5db0f', ('.0',)), ('eq-55ca3c29ae249bb1a1ab9eb42332a747', ('.0',))) -> getitem-56f4e42f7c46ae412827cce188b37184>,
 'read-csv-136524d4e292271449abaac2e2d5db0f': DataFrameIOLayer<name='read-csv-136524d4e292271449abaac2e2d5db0f', n_parts=4, columns=['timestamp', 'id', 'name', 'x', 'y']>,
 'getitem-063bc8eb2c47ea9536bbd7059706a728': Blockwise<(('read-csv-136524d4e292271449abaac2e2d5db0f', ('.0',)), ('name', None)) -> getitem-063bc8eb2c47ea9536bbd7059706a728>,
 'eq-55ca3c29ae249bb1a1ab9eb42332a747': Blockwise<(('getitem-063bc8eb2c47ea9536bbd7059706a728', ('.0',)), ('Alice', None)) -> eq-55ca3c29ae249bb1a1ab9eb42332a747>}
```

**Comment by GenevieveBuckley** [comment]
For the layers specifically, I used the `toposort` function to re-order them for the high level graph text representation.

Are there lots of other places besides layers where this is noticeable?

**Comment by mrocklin** [comment]
Heads up, some toposort implementations are also not deterministic.  There
are many valid toposorts.

**Comment by mrocklin** [comment]
More generally, I wouldn't want us to use a different data structure just
for doctests.  However in some cases, like the HLG case above, it probably
does make sense for us to be consistent.

With regards to unskipping doctests I wouldn't go too crazy here.  In some
cases it's perfectly fine to skip doctests.  I would unskip them with
intention, rather than in a blanket effort.

**Comment by jsignell** [comment]
> More generally, I wouldn't want us to use a different data structure just
for doctests.  However in some cases, like the HLG case above, it probably
does make sense for us to be consistent.

Oh yeah, I'm definitely not saying that we should do this if the only benefit is that the doctests run. But I do think that in this case the doctests have exposed an area that could be improved (which is hopefully what tests are for!).

> With regards to unskipping doctests I wouldn't go too crazy here.  In some
cases it's perfectly fine to skip doctests.  I would unskip them with
intention, rather than in a blanket effort.

I tend to disagree. When doctests are skipped they tend to get out of date or be incomplete examples and that can make things confusing for users. I think that skipping doctests should be the exception and people should be pushed to try to write runnable examples.

**Comment by mrocklin** [comment]
> I tend to disagree. When doctests are skipped they tend to get out of
date or be incomplete examples and that can make things confusing for
users. I think that skipping doctests should be the exception and people
should be pushed to try to write runnable examples.

I agree.  But in many cases we skip things intentionally because they don't
make sense to run as doctests.  I wouldn't want a blind effort to unskip
doctests.

**Comment by jsignell** [comment]
Yeah I agree, not a blind effort, but maybe a good-faith attempt to unskip. I think there have been some doctests where the doctests are just skipped because of copy/pasting from other places in the code or maybe not knowing about some of the othe doctest comment options. @zzhengnan is working on this with what feels to me like a very sane approach.

**Comment by mrocklin** [comment]
OK.  Grand
